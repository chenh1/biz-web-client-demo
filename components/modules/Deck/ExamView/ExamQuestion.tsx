'use client'

import type { ExamQuestionType } from '@/graphql/queries/examQuestions';
import { Paragraph } from '../../../core/Paragraph/Paragraph';
import { Box } from '../../../containers/Box/Box';
import React, { useState } from 'react';
import { Button } from '../../../core/Button/Button';
import { ClickMultiChoice } from './ClickMultiChoice';
import { InputFillInBlank } from './InputFillInBlank';
import { InteractiveTable } from './InteractiveTable';
import { ClickInteractiveChart } from './ClickInteractiveChart';


interface ExamQuestionProps {
  question: ExamQuestionType;
  invokeNextQuestion: () => void;
  incrementCorrectAnswers: () => void;
}

export const ExamQuestion: React.FC<ExamQuestionProps> = ({ question, invokeNextQuestion, incrementCorrectAnswers }: ExamQuestionProps) => {
  const { multiChoiceQuestion, fillInQuestion, interactiveTableQuestion, interactiveChartQuestion } = question;
  const questionType = !!multiChoiceQuestion
    ? 'multi_choice'
    : !!fillInQuestion
      ? 'fill_in'
      : !!interactiveTableQuestion
        ? 'interactive_table'
        : !!interactiveChartQuestion
          ? 'interactive_chart'
          : null;

  const [ isCorrect, setIsCorrect ] = useState<boolean>(false);
  const [ userInputAnswerSet, updateUserInputAnswerSetTo ] = useState([]); // for multi choice questions
  const [ isAnswered, setIsAnswered ] = useState<boolean>(false);

  // base compare answer
  const compareAnswer = (answer: any, value: any): boolean => {
    return typeof answer === 'number'
      ? answer === parseFloat(value)
      : typeof answer === 'string'
        ? answer.toLowerCase?.() === value.toLowerCase?.()
        : answer === value;
  }

  // batch compare answer for multiple correct answers or multiple fill in the blanks
  const compareAnswers = (answers, values, requiresOnlyOne) => {
    const correctAnswers = answers?.map((answer, i) => compareAnswer(answer, values[i]));
    updateUserInputAnswerSetTo(correctAnswers)
    
    if (requiresOnlyOne) {
      return correctAnswers?.some(answer => answer === true);
    }

    return correctAnswers?.every(answer => answer === true);
  }


  const onAnswer = (value) => {
    setIsAnswered(true);
    let isCorrect = false;

    if (questionType === 'multi_choice') { // value will be the answer key
      isCorrect = question?.multiChoiceQuestion?.correctAnswers?.find(answer => compareAnswer(answer.key, value)) !== undefined;
    }

    if (questionType === 'fill_in') { // value will be the input value
      isCorrect = compareAnswers(question?.fillInQuestion?.correctAnswers?.map(answer => answer.answer), value, false);
    }

    if (questionType === 'interactive_table') {
      const answerType = question?.interactiveTableQuestion?.answerType;

      if (answerType === 'one_click') {
        isCorrect = compareAnswers(question?.interactiveTableQuestion?.correctAnswers?.map(answer => answer.value), [ value ], true);
      } else {
        isCorrect = compareAnswers(question?.interactiveTableQuestion?.correctAnswers?.map(answer => answer.value), value, false);
      }
    }

    if (questionType === 'interactive_chart') {
      const chartType = question?.interactiveChartQuestion?.chartType;

      if (chartType === 'candlestick') {
        isCorrect = compareAnswer(question?.interactiveChartQuestion?.correctAnswer?.x, value?.x);
      } else {
        const xIsCorrect = compareAnswer(question?.interactiveChartQuestion?.correctAnswer?.x, value?.x);
        const yIsCorrect = compareAnswer(question?.interactiveChartQuestion?.correctAnswer?.y, value?.y);
        isCorrect = xIsCorrect && yIsCorrect;
      }
    }

    setIsCorrect(!!isCorrect);

    if (isCorrect) {
      incrementCorrectAnswers();
    }
  }
  
  const renderQuestion = () => {
    if (questionType === 'multi_choice' && !!multiChoiceQuestion) {
      return (
        // multi choice
        <ClickMultiChoice isAnswered={isAnswered} question={multiChoiceQuestion} onAnswer={onAnswer} />
      )
    } else if (questionType === 'fill_in' && !!fillInQuestion) {
      return (
        // fill in answer
        <InputFillInBlank question={fillInQuestion} userInputAnswerSet={userInputAnswerSet} onAnswer={onAnswer} />
      )
    } else if (questionType === 'interactive_table' && !!interactiveTableQuestion) {
      return (
        // interactive
        <InteractiveTable isAnswered={isAnswered} question={interactiveTableQuestion} onAnswer={onAnswer} />
      )
    } else if (questionType === 'interactive_chart' && !!interactiveChartQuestion) {
      return (
        // interactive
        <ClickInteractiveChart isAnswered={isAnswered} question={interactiveChartQuestion} onAnswer={onAnswer} />
      )
    } else {
      return (
        <></>
      )
    }
  }



  return (
    <Box padding="0">
      {renderQuestion()}
      {isAnswered &&
        <Box padding="0" className="max-w-xl">
          <Paragraph>{isCorrect ? 'Correct!' : 'Incorrect...'}</Paragraph>
          <Paragraph>{question?.explanation}</Paragraph>
          <Button label="Next" primary onClick={() => {
            //scroll to top of window whenever going to next question
            window.scrollTo(0, 0);

            setIsAnswered(false);
            setIsCorrect(false);
            updateUserInputAnswerSetTo([])
            invokeNextQuestion()
          }}/>
        </Box>
      }
    </Box>
  )
}