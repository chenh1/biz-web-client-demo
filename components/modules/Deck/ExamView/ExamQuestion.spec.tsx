import { ExamQuestion } from "./ExamQuestion";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("@visx/responsive", () => ({
  ParentSize: ({ children }: { children: (size: { width: number, height: number }) => JSX.Element }) => children({ width: 100, height: 100 })
}))

// mock LocalResizeObserver
global.LocalResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}))


it('renders an exam question with multiple choice', async () => {
  const { getByTestId, asFragment } = render(
    <ExamQuestion
      invokeNextQuestion={() => {}}
      incrementCorrectAnswers={() => {}}
      question={{
        explanation: 'Explanation',
        lessonModule: 1.01,
        tags: {
          data: {
            attributes: {
              displayName: 'Tag',
              color: 'red',
            },
          },
        },
        multiChoiceQuestion: {
          question: 'What is 1 + 1?',
          answers: [
            { key: '2', answer: '2', value: '2' },
            { key: '3', answer: '3', value: '3' },
            { key: '4', answer: '4', value: '4' },
            { key: '5', answer: '5', value: '5' },
          ],
          correctAnswers: [{ key: '2', answer: '2', value: '2' }],
        }
      }}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders an exam question with a fill-in field', async () => {
  const { getByTestId, asFragment } = render(
    <ExamQuestion
      invokeNextQuestion={() => {}}
      incrementCorrectAnswers={() => {}}
      question={{
        explanation: 'Explanation',
        lessonModule: 1.01,
        tags: {
          data: {
            attributes: {
              displayName: 'Tag',
              color: 'red',
            },
          },
        },
        fillInQuestion: {
          question: 'What is 1 + 1?',
          correctAnswers: [{ key: '2', answer: '2', value: '2' }],
        },
      }}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders an exam question with an interactive table', async () => {
  const { getByTestId, asFragment } = render(
    <ExamQuestion
      invokeNextQuestion={() => {}}
      incrementCorrectAnswers={() => {}}
      question={{
        explanation: 'Explanation',
        lessonModule: 1.01,
        tags: {
          data: {
            attributes: {
              displayName: 'Tag',
              color: 'red',
            },
          },
        },
        interactiveTableQuestion: {
          question: 'What is 1 + 1?',
          answerType: 'number',
          tableData: [
            { value: '2', xCell: 0, yCell: 0 },
            { value: '3', xCell: 1, yCell: 0 },
            { value: '4', xCell: 2, yCell: 0 },
            { value: '5', xCell: 3, yCell: 0 },
          ],
          answers: [{ key: '2', answer: '2', value: '2', xCell: 0, yCell: 0 }],
          correctAnswers: [{ key: '2', answer: '2', value: '2', xCell: 0, yCell: 0 }],
          rowHeaders: [{ header: 'Row 1', position: 0 }],
          columnHeaders: [{ header: 'Column 1', position: 0 }],
        },
      }}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

// it('renders an exam question with an interactive chart', async () => {
//   const { getByTestId, asFragment } = render(
//     <ExamQuestion
//       invokeNextQuestion={() => {}}
//       incrementCorrectAnswers={() => {}}
//       question={{
//         explanation: 'Explanation',
//         lessonModule: 1.01,
//         tags: {
//           data: {
//             attributes: {
//               displayName: 'Tag',
//               color: 'red',
//             },
//           },
//         },
//         interactiveChartQuestion: {
//           xType: 'number',
//           yType: 'number',
//           xAxisLabel: 'X Axis',
//           yAxisLabel: 'Y Axis',
//           question: 'What is 1 + 1?',
//           chartType: 'line',
//           chartData: [
//             { x: 1, y: 1, multiPoints: { open: 1, close: 1, low: 1, high: 1 }},
//             { x: 2, y: 2, multiPoints: { open: 2, close: 2, low: 2, high: 2 }},
//             { x: 3, y: 3, multiPoints: { open: 3, close: 3, low: 3, high: 3 }},
//           ],
//           correctAnswers: [{ x: 2, y: 2, key: '2', answer: '2' }],
//           correctAnswer: { x: 2, y: 2, key: '2', answer: '2' },
//           interactiveChartData: []
//         },
//       }}
//     />
//   );

//   expect(asFragment()).toMatchSnapshot();
// });

it('renders an empty wrapper if no question data is passed in', async () => {
  const { getByTestId, asFragment } = render(
    <ExamQuestion
      invokeNextQuestion={() => {}}
      incrementCorrectAnswers={() => {}}
      question={{
        explanation: 'Explanation',
        lessonModule: 1.01,
        tags: {
          data: {
            attributes: {
              displayName: 'Tag',
              color: 'red',
            },
          },
        }
      }}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});
