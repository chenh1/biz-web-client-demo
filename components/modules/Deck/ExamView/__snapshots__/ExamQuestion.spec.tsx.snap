// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty wrapper if no question data is passed in 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 "
  />
</DocumentFragment>
`;

exports[`renders an exam question with a fill-in field 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 "
  >
    <div
      class=" grid  gap-4  p-2 justify-center"
    >
      <div
        class=" grid  gap-8  p-2 max-w-xl"
      >
        <div
          class=" grid  gap-4  p-2 "
        >
          <form>
            <p
              class="inline-block text-night-black dark:text-max-white font-body font-body text-base mr-1"
            >
              What is 1 + 1?
            </p>
            <button
              class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5   "
              data-testid="button"
              type="submit"
            >
              <div
                class="triangle left"
              >
                <div />
              </div>
              <div
                class="triangle right"
              >
                <div />
              </div>
              <span
                class="pointer-events-none relative z-10"
              >
                Submit
              </span>
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders an exam question with an interactive table 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 "
  >
    <div
      class=" grid  gap-4  p-2 justify-center"
    >
      <div
        class=" grid  gap-8  p-2 max-w-xl"
      >
        <h4
          class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md "
        >
          What is 1 + 1?
        </h4>
        <form>
          <div
            class=" grid  gap-8  p-2 "
          >
            <div
              class=" grid  gap-4  p-2 relative max-w-full"
            >
              <div
                class=" grid  gap-4  p-2 fade-scroller max-w-full overflow-x-auto overflow-y-hidden"
              >
                <div
                  class="grid gap-4 min-w-[600px]"
                  style="grid-template-columns: repeat(5, minmax(0, 1fr));"
                >
                  <div />
                  <div
                    cellindex="1"
                    class=" grid rounded-md gap-4 shadow-lg p-8 "
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                  >
                    <h4
                      class="text-night-black dark:text-max-white font-headings font-semibold text-md "
                    >
                      Column 1
                    </h4>
                  </div>
                  <div
                    cellindex="0"
                    class=" grid rounded-md gap-4 shadow-lg p-8 "
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                  >
                    <h4
                      class="text-night-black dark:text-max-white font-headings font-semibold text-md "
                    >
                      Row 1
                    </h4>
                  </div>
                  <div
                    cellindex="0"
                    class=" grid rounded-md gap-4 shadow-lg p-8 cursor-pointer bg-night-black-100 hover:bg-night-black-200"
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                    value="2"
                  >
                    <p
                      class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
                    >
                      2
                    </p>
                  </div>
                  <div
                    cellindex="1"
                    class=" grid rounded-md gap-4 shadow-lg p-8 cursor-pointer bg-night-black-100 hover:bg-night-black-200"
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                    value="3"
                  >
                    <p
                      class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
                    >
                      3
                    </p>
                  </div>
                  <div
                    cellindex="2"
                    class=" grid rounded-md gap-4 shadow-lg p-8 cursor-pointer bg-night-black-100 hover:bg-night-black-200"
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                    value="4"
                  >
                    <p
                      class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
                    >
                      4
                    </p>
                  </div>
                  <div
                    cellindex="3"
                    class=" grid rounded-md gap-4 shadow-lg p-8 cursor-pointer bg-night-black-100 hover:bg-night-black-200"
                    columnheaders="[object Object]"
                    numberofcellsincolumn="1"
                    numberofcellsinrow="4"
                    rowheaders="[object Object]"
                    rowindex="0"
                    value="5"
                  >
                    <p
                      class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
                    >
                      5
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders an exam question with multiple choice 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 "
  >
    <div
      class=" grid  gap-4  p-2 justify-center"
    >
      <div
        class=" grid  gap-8  p-2 max-w-xl"
      >
        <h4
          class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md "
        >
          What is 1 + 1?
        </h4>
        <div
          class=" grid  gap-4  p-2 "
        >
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5 text-left  "
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              2
            </span>
          </button>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5 text-left  "
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              3
            </span>
          </button>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5 text-left  "
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              4
            </span>
          </button>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5 text-left  "
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              5
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
